src/bootstrap/builder.rs

```diff
@@ -284,18 +284,25 @@ impl StepDescription {
         }
     }
 
-    fn maybe_run(&self, builder: &Builder<'_>, pathsets: Vec<PathSet>) {
+    /// This might run the builder. Returns false if no targets can be determined indicating that
+    /// builder was not run
+    fn maybe_run(&self, builder: &Builder<'_>, pathsets: Vec<PathSet>) -> bool {
         if pathsets.iter().any(|set| self.is_excluded(builder, set)) {
-            return;
+            return true;
         }
 
         // Determine the targets participating in this rule.
         let targets = if self.only_hosts { &builder.hosts } else { &builder.targets };
+        if targets.is_empty() && self.only_hosts {
```

Comment:
This condition doesn't seem to match the comment above - if targets is empty, regardless of only host we won't execute any steps, right? So it's not clear to me exactly what the if here is trying to do.

Reply:
If hosts is empty, the current way is to revert to targets. So I am checking if we didn't revert to target and targets is empty. It is case 3 that I mention here: https://github.com/rust-lang/rust/pull/107858#discussion_r1103871008

Reply:
> If hosts is empty, the current way is to revert to targets.

Is this true? It seems to me that we *always* [default targets](https://github.com/rust-lang/rust/blob/master/src/bootstrap/config.rs#L1015-L1023) to the `hosts` list in config.rs unless there's an explicit target override.

If the goal is to check the condition you're interested in, then I would adjust the previous if statement:

```
let targets = if self.only_hosts {
    if builder.hosts.is_empty() {
         return false;
    }
    &builder.hosts
} else {
    &builder.targets
};
```

But I remain confused by the goal in checking this condition *here* rather than in src/bootstrap/config.rs, regardless of which specific step is being run...